pkg/osquery/log_publication_state.go

```diff
@@ -0,0 +1,105 @@
+	// maxBytesPerBatch is passed in from the extension opts and respected
+	// as a fixed upper limit for batch size, regardless of publication success/failure rates
+	maxBytesPerBatch int
+	// currentMaxBytesPerBatch represents the (stateful) upper limit being enforced
+	currentMaxBytesPerBatch int
+	// currentBatchBufferFilled is used to indicate when a batch's success can be used
+	// to increase the threshold (we only want to increase the threshold after sending full, not partial, batches successfully)
+	currentBatchBufferFilled bool
+	currentBatchStartTime    time.Time
+}
+
+func NewLogPublicationState(maxBytesPerBatch int) *logPublicationState {
+	return &logPublicationState{
+		maxBytesPerBatch:        maxBytesPerBatch,
+		currentMaxBytesPerBatch: maxBytesPerBatch,
+	}
+}
+
+// BeginBatch sets the opening state before attempting to publish a batch of logs. Specifically, it must
+// - note the time (to determine if an error later was timeout related)
+// - note whether this batch is full (to determine whether success should increase the limit on success later)
+func (lps *logPublicationState) BeginBatch(startTime time.Time, bufferFilled bool) {
+	lps.currentBatchStartTime = startTime
+	lps.currentBatchBufferFilled = bufferFilled
+}
+
+func (lps *logPublicationState) CurrentValues() map[string]int {
+	return map[string]int{
+		"options_batch_limit_bytes": lps.maxBytesPerBatch,
+		"current_batch_limit_bytes": lps.currentMaxBytesPerBatch,
+	}
+}
+
+func (lps *logPublicationState) EndBatch(logs []string, successful bool) {
+	// ensure we reset all batch state at the end
+	defer func() {
+		lps.currentBatchBufferFilled = false
+		lps.currentBatchStartTime = time.Time{}
+	}()
+
+	// we can always safely decrease the threshold for a failed batch, but
+	// shouldn't increase the threshold for a successful batch unless we've at
+	// least filled the buffer
+	if successful && !lps.currentBatchBufferFilled {
+		return
+	}
+
+	// in practice there could be one of a few different transport or timeout errors that bubble up
+	// depending on network conditions. instead of trying to keep up with all potential errors,
+	// only reduce the threshold if the calls are failing after more than 20 seconds
+	if !successful && time.Since(lps.currentBatchStartTime) < maxPublicationDuration {
+		return
+	}
+
+	if successful {
+		lps.increaseBatchThreshold()
+		return
+	}
+
+	lps.reduceBatchThreshold()
+}
+
+func (lps *logPublicationState) ExceedsCurrentBatchThreshold(amountBytes int) bool {
+	return amountBytes > lps.currentMaxBytesPerBatch
+}
+
+func (lps *logPublicationState) reduceBatchThreshold() {
+	if lps.currentMaxBytesPerBatch <= minBytesPerBatch {
+		return
+	}
+
+	newTargetThreshold := lps.currentMaxBytesPerBatch - batchIncrementAmount
+	lps.currentMaxBytesPerBatch = maxInt(newTargetThreshold, minBytesPerBatch)
+}
+
+func (lps *logPublicationState) increaseBatchThreshold() {
+	if lps.currentMaxBytesPerBatch >= lps.maxBytesPerBatch {
+		return
+	}
```

Comment:
Nice. I _suspect_ we'll need to figure out how to grow beyond this -- some customers have it coded really small. But maybe this is fine to start, and we can make a real max value later